src/fastpath/fastpath.c

```diff
@@ -15,6 +15,268 @@
 #endif
 #include <benchmark/benchmark_utilisation.h>
 
+#ifdef CONFIG_ARCH_ARM
+static inline
+#ifndef CONFIG_ARCH_ARM_V6
+FORCE_INLINE
+#endif
+#endif
+void NORETURN fastpath_signal(word_t cptr, word_t msgInfo)
+{
+    cap_t newVTable;
+    vspace_root_t *cap_pd;
+    pde_t stored_hw_asid;
+    word_t fault_type;
+    dom_t dom;
+
+    /* Get fault type. */
+    fault_type = seL4_Fault_get_seL4_FaultType(NODE_STATE(ksCurThread)->tcbFault);
+
+    /* Check there's no saved fault */
+    if (unlikely(fault_type != seL4_Fault_NullFault)) {
+        slowpath(SysSend);
+    }
+
+    /* We land up in here on every SYSCALL_SEND, so we need to check that we are
+     * actually signalling a notification */
```

Comment:
Can this be checked in the assembly before the call to this function?

Reply:
As in, can we check if we're signalling a notification (rather than sending a message to an endpoint) in the assembly?

As far as I know the way that we distinguish between whether a syscall is signalling a notification or sending to an endpoint is by checking the cap type (via `cap_capType_equals(cap, cap_notification_cap)`), since both syscalls use the `SYSCALL_SEND` code.

I think if we added a separate code for signalling a notification (e.g. `SYSCALL_SIGNAL`), then we could check in the assembly code whether we're signalling a notification. Would that be feasible?

Reply:
I don't think so, that would require changes to the syscall interface. Would need to discuss with others on that. @xurtis thoughts?

Reply:
I think it's worth exploring adding a separate sycall code. @Xaphiosis can you comment on what effect that would that have proof wise?

Reply:
That does not sound like a good idea -- not so much for the proof (although it would lead to duplication), but more because it would break the general concept that everything is a `send` to a cap.

Don't we have the same problem in the fast path for normal endpoints? What do we do there?

Reply:
This might require a bit more thinking before even hitting verification. seL4 seems to have a design goal of not having many system calls, and selecting its actions by cap type instead. You'd have to introduce the new system call changing the syscall datatype, then move decodes and performs around... it probably will be a bunch of non-interesting work, spec shuffling, and result in some proof duplication. Then there's updating the security property proofs, capDL, and all the other proofs on the other verified architectures. I don't expect anything *exciting* to come up, but I've been wrong before.
For doing it on master, there might be no one available to deal with it in the near future.
As this looks MCS only, the people on that verification project would no doubt want to have a say. @corlewis is handling the verification team and can say more.

It would be nice to look at a prototype and diff in C that does what people actually want and evaluate verification cost from that before embarking on proof adventures.

Usual warning for fastpaths and changes: the most common issue we find during verification is that people think their fastpath does what the slow path does in the same circumstances, but it turns out to not be the case.

Reply:
Thanks, that gives something to think on.

Reply:
> Don't we have the same problem in the fast path for normal endpoints? What do we do there?

I tried to use the same approach as what we do in [`fastpath_call`](https://github.com/seL4/seL4/blob/master/src/fastpath/fastpath.c#L51-L57), which is to lookup the cap and check that it's an endpoint (but in this case we check that it's a notification).

Reply:
Let me know if this is okay to mark as resolved.